### ISSUE TITLE
New API to localize next generation components

### ISSUE THREAD

**Issue by web-padawan** [vaadin/vaadin-core#237]
Current generation has a serious problem described at vaadin/vaadin-upload-flow#71

> The example should show how you define the generic translations for all Upload components, who wants to translate a single upload dynamically?

We should provide a new API that would cover at least the following components:

- [ ] vaadin-date-picker
- [ ] vaadin-upload
- [ ] vaadin-time-picker
- [ ] vaadin-rich-text-editor

Take a look at https://github.com/ing-bank/lion/tree/master/packages/localize for inspiration:

- get locale from `lang` attribute on `<html>`, same as we currently do for `dir` ([example](https://github.com/ing-bank/lion/blob/bc7c3f9216c0fa834b37c9103f500336290e97f4/packages/localize/src/LocalizeManager.js#L35))
- use `Intl` where possible, e.g. we can use [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat) to get [weekdays](https://github.com/ing-bank/lion/blob/bc7c3f9216c0fa834b37c9103f500336290e97f4/packages/localize/src/date/getWeekdayNames.js#L46) / [month names](https://github.com/ing-bank/lion/blob/bc7c3f9216c0fa834b37c9103f500336290e97f4/packages/localize/src/date/getMonthNames.js#L11)

**Comment by jouni** [comment]
Probably obvious, but whatever we implement, it should be usable in custom components as well. Also, developers should be able to “extend” the localization messages, so that they can use the same message easily across an app for example.

This might also be trivial/irrelevant to point out, but remember components like Confirm Dialog and CRUD, which should also offer some default UI texts. Some of those might be generic enough that developers could use them in some other places in their apps (e.g. “Are you sure you want to delete {item}?”). The idea I’ve had in my personal notes is “Packaged localized ui texts / microcopy: share i18n texts across components”.

**Comment by web-padawan** [comment]
> developers should be able to “extend” the localization messages, so that they can use the same message easily across an app for example

Yes that's part of the idea. In particular, `clear` button is used in many components and should only have a single localisation string.

We can look at [LocalizeMixin](https://github.com/ing-bank/lion/blob/master/packages/localize/src/LocalizeMixin.js) from Lion for a solution.

**Comment by web-padawan** [comment]
One more use case for new API would be to allow re-using translations across components.

See https://github.com/vaadin/vaadin-combo-box/issues/868 and https://github.com/vaadin/vaadin-select/issues/218 as an example.
